fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DbFedoraToOcflObjectIndex.java

```diff
@@ -54,67 +70,258 @@
+
+    /**
+     * Map of database product name to COMMIT to mapping table from operations table
+     */
+    private static final Map<String, String> COMMIT_ADD_MAPPING_MAP = Map.of(
+            "MySQL", COMMIT_ADD_MAPPING_MYSQL_MARIA,
+            "H2", COMMIT_ADD_MAPPING_H2,
+            "PostgreSQL", COMMIT_ADD_MAPPING_POSTGRESQL,
+            "MariaDB", COMMIT_ADD_MAPPING_MYSQL_MARIA
+    );
 
-    private static final String REMOVE_MAPPING = "DELETE FROM " + MAPPING_TABLE +
-            " WHERE " + FEDORA_ID_COLUMN + " = :fedoraId";
+    /*
+     * Delete records from the mapping table that are to be deleted in this transaction.
+     */
+    private static final String COMMIT_DELETE_RECORDS = "DELETE FROM " + MAPPING_TABLE + " WHERE " +
+            "EXISTS (SELECT * FROM " + TRANSACTION_OPERATIONS_TABLE + " WHERE " +
+            TRANSACTION_ID_COLUMN + " = :transactionId AND " +  OPERATION_COLUMN + " = 'delete' AND " +
+            MAPPING_TABLE + "." + FEDORA_ID_COLUMN + " = " + TRANSACTION_OPERATIONS_TABLE + "." + FEDORA_ID_COLUMN +
+            ")";
 
     private static final String TRUNCATE_MAPPINGS = "TRUNCATE TABLE " + MAPPING_TABLE;
 
+    private static final String TRUNCATE_TRANSACTIONS = "TRUNCATE TABLE " + TRANSACTION_OPERATIONS_TABLE;
+
+    /*
+     * Delete all records from the transaction table for the specified transaction.
+     */
+    private static final String DELETE_ENTIRE_TRANSACTION = "DELETE FROM " + TRANSACTION_OPERATIONS_TABLE + " WHERE " +
+            TRANSACTION_ID_COLUMN + " = :transactionId";
+
+    /*
+     * Row mapper for the Lookup queries.
+     */
+    private static final RowMapper<FedoraOCFLMapping> GET_MAPPING_ROW_MAPPER = (resultSet, i) -> new FedoraOCFLMapping(
+            resultSet.getString(1),
+            resultSet.getString(2)
+    );
+
     private final DataSource dataSource;
 
     private final NamedParameterJdbcTemplate jdbcTemplate;
 
+    private PlatformTransactionManager platformTransactionManager;
+
+    private String databaseProductName;
+
     public DbFedoraToOcflObjectIndex(@Autowired final DataSource dataSource) {
         this.dataSource = dataSource;
+        this.platformTransactionManager = new DataSourceTransactionManager(dataSource);
         this.jdbcTemplate = new NamedParameterJdbcTemplate(dataSource);
     }
 
     @PostConstruct
     public void setup() {
-        LOGGER.info("Applying ddl: {}", DDL);
+        final var ddl = lookupDdl();
+        LOGGER.info("Applying ddl: {}", ddl);
         DatabasePopulatorUtils.execute(
-                new ResourceDatabasePopulator(new DefaultResourceLoader().getResource("classpath:" + DDL)),
+                new ResourceDatabasePopulator(new DefaultResourceLoader().getResource("classpath:" + ddl)),
                 dataSource);
     }
 
+    private String lookupDdl() {
+        try (final var connection = dataSource.getConnection()) {
```

Comment:
Ideally, the db platform names would be an enum like:

```java
public enum DbPlatform {
    POSTGRESQL("PostgreSQL"), H2("H2"), MYSQL("MySQL"), MARIADB("MariaDB");

    private final String name;

    private DbPlatform(final String name) {
        this.name = name;
    }

    private static DbPlatform fromDataSource(final DataSource dataSource) {
        try (final var connection = dataSource.getConnection()) {
            return fromString(connection.getMetaData().getDatabaseProductName());
        } catch (final SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private static DbPlatform fromString(final String name) {
        for (var value : values()) {
            if (value.name.equals(name)) {
                return value;
            }
        }
        throw new IllegalArgumentException("Unknown database platform: " + name);
    }
}
```

But, it's unclear to me if there is a good common library to put that code in.

At the minimum, setting the `databaseProductName` needs to happen in the `setup()` method and not as a side effect of calling `lookupDdl`.

Reply:
If there isn't a good place to share the code (I believe I have duplicate code in another class). Leaving the duplicate, string based code is fine for now.